LoRaEngine/modules/LoRaWanNetworkSrvModule/LoRaWan.NetworkServer/LoRaDeviceClientConnectionManager.cs

```diff
@@ -1,162 +1,399 @@
+                    {
+                        var ((id, name), outcome) = args;
+                        someLogger.LogDebug(@"Invoked ""{Name}"" ({Id}); status = {Status}, run-time = {RunTime}, wait-time = {WaitTime}",
+                                            name, id, outcome.Task.Status, outcome.RunDuration, outcome.WaitDuration);
+                    };
+
+                    this.exclusiveProcessor.Interrupted += (_, p) =>
+                    {
+                        var (interrupted, interrupting) = p;
+                        someLogger.LogDebug(@"Interrupted ""{Name}"" ({Id}) by {InterruptingName} ({InterruptingId})",
+                                            interrupted.Name, interrupted.Id,
+                                            interrupting.Name, interrupting.Id);
+                    };
+                }
             }
-        }
 
-        private static string GetConnectionCacheKey(DevEui devEui) => string.Concat("connection:", devEui);
+            public DevEui DevEui => this.device.DevEUI;
+            public TimeSpan KeepAliveTimeout => TimeSpan.FromSeconds(this.device.KeepAliveTimeout);
 
-        private static string GetScheduleCacheKey(DevEui devEui) => string.Concat("connection:schedule:", devEui);
+            public void Dispose() => this.client?.Dispose();
 
-        public ILoRaDeviceClient GetClient(LoRaDevice loRaDevice)
-        {
-            if (loRaDevice is null) throw new ArgumentNullException(nameof(loRaDevice));
+            public event EventHandler? EnsureConnectedSucceeded;
 
-            return this.managedConnections.TryGetValue(GetConnectionCacheKey(loRaDevice.DevEUI), out var managedConnection)
-                 ? managedConnection.DeviceClient
-                 : throw new ManagedConnectionException($"Connection for device {loRaDevice.DevEUI} was not found");
-        }
+            private Task<T> InvokeExclusivelyAsync<T>(Func<ILoRaDeviceClient, Task<T>> processor,
+                                                      [CallerMemberName] string? callerName = null) =>
+                InvokeExclusivelyAsync(doesNotRequireOpenConnection: false, processor, callerName);
 
-        public void Register(LoRaDevice loRaDevice, ILoRaDeviceClient loraDeviceClient)
-        {
-            if (loRaDevice is null) throw new ArgumentNullException(nameof(loRaDevice));
+            private async Task<T> InvokeExclusivelyAsync<T>(bool doesNotRequireOpenConnection,
+                                                            Func<ILoRaDeviceClient, Task<T>> processor,
+                                                            [CallerMemberName] string? callerName = null)
+            {
+                _ = Interlocked.Increment(ref this.pid);
+                return await this.exclusiveProcessor.ProcessAsync(new Process(this.pid, callerName!), async () =>
+                {
+                    if (!doesNotRequireOpenConnection)
```

Comment:
nit: if you change this to "requiresOpenConnection" it becomes more readable.

Reply:
is this ever needed? what operation would not require an open connection?

Reply:
> is this ever needed? what operation would not require an open connection?

`DisconnectAsync`

Reply:
> nit: if you change this to "requiresOpenConnection" it becomes more readable.

I know what you mean, but a recommended practice that I am following here is that an optional `Boolean` value should default to `false` otherwise it gets very hard to remember (without consulting the source) what's the default behaviour. It does indeed sometimes mean double-negation from a naming/usage perspective.